pkg/services/ngalert/schedule/metrics.go

```diff
@@ -56,20 +72,59 @@ func (sch *schedule) updateRulesMetrics(alertRules []*models.AlertRule) {
 		orgGroups[rule.RuleGroup] = struct{}{}
 	}
 
-	for orgID, numRules := range rulesPerOrg {
-		numRulesPaused := orgsPaused[orgID]
-		numRulesNfSettings := orgsNfSettings[orgID]
-		sch.metrics.GroupRules.WithLabelValues(fmt.Sprint(orgID), metrics.AlertRuleActiveLabelValue).Set(float64(numRules - numRulesPaused))
-		sch.metrics.GroupRules.WithLabelValues(fmt.Sprint(orgID), metrics.AlertRulePausedLabelValue).Set(float64(numRulesPaused))
-		sch.metrics.SimpleNotificationRules.WithLabelValues(fmt.Sprint(orgID)).Set(float64(numRulesNfSettings))
+	for key, numRules := range rulesPerOrgFolderGroup {
+		numRulesPaused := rulesPerOrgFolderGroupPaused[key]
+		ruleGroupLabelValue := makeRuleGroupLabelValue(key)
+		sch.metrics.GroupRules.WithLabelValues(fmt.Sprint(key.OrgID), metrics.AlertRuleActiveLabelValue, ruleGroupLabelValue).Set(float64(numRules - numRulesPaused))
+		sch.metrics.GroupRules.WithLabelValues(fmt.Sprint(key.OrgID), metrics.AlertRulePausedLabelValue, ruleGroupLabelValue).Set(float64(numRulesPaused))
 	}
 
-	for orgID, groups := range groupsPerOrg {
-		sch.metrics.Groups.WithLabelValues(fmt.Sprint(orgID)).Set(float64(len(groups)))
+	for orgID := range updateMetricsForOrgsAndGroups {
+		sch.metrics.SimpleNotificationRules.WithLabelValues(fmt.Sprint(orgID)).Set(float64(orgsNfSettings[orgID]))
+		sch.metrics.Groups.WithLabelValues(fmt.Sprint(orgID)).Set(float64(len(groupsPerOrg[orgID])))
 	}
 
 	// While these are the rules that we iterate over, at the moment there's no 100% guarantee that they'll be
 	// scheduled as rules could be removed before we get a chance to evaluate them.
 	sch.metrics.SchedulableAlertRules.Set(float64(len(alertRules)))
 	sch.metrics.SchedulableAlertRulesHash.Set(float64(hashUIDs(alertRules)))
+
+	// Delete metrics for rule groups and orgs that are no longer present in the new state
+	for orgID, alertRuleGroupsMap := range sch.lastUpdatedMetricsForOrgsAndGroups {
+		if orgOrGroupDeleted(updateMetricsForOrgsAndGroups, orgID, nil) {
+			sch.metrics.SimpleNotificationRules.DeleteLabelValues(fmt.Sprint(orgID))
+			sch.metrics.Groups.DeleteLabelValues(fmt.Sprint(orgID))
+		}
+
+		for key := range alertRuleGroupsMap {
+			if orgOrGroupDeleted(updateMetricsForOrgsAndGroups, orgID, &key) {
+				ruleGroupLabelValue := makeRuleGroupLabelValue(key)
+				sch.metrics.GroupRules.DeleteLabelValues(fmt.Sprint(key.AlertRuleGroupKey.OrgID), metrics.AlertRuleActiveLabelValue, ruleGroupLabelValue)
+				sch.metrics.GroupRules.DeleteLabelValues(fmt.Sprint(key.AlertRuleGroupKey.OrgID), metrics.AlertRulePausedLabelValue, ruleGroupLabelValue)
+			}
+		}
+	}
+
+	// update the call state
+	sch.lastUpdatedMetricsForOrgsAndGroups = updateMetricsForOrgsAndGroups
+}
+
+// makeRuleGroupLabelValue returns a string that can be used as a label (rule_group) value for alert rule group metrics.
+func makeRuleGroupLabelValue(key models.AlertRuleGroupWithFolderTitleKey) string {
+	return fmt.Sprintf("%s;%s", key.FolderTitle, key.AlertRuleGroupKey.RuleGroup)
```

Comment:
I am not sure the folder title makes sense here:
- it can contain up to 255 characters
- it can change when a user changes the title.

I think namespace UID would be better here:
- it has a fixed length
- it is mentioned in logs 
- it is used in other places, e.g. RBAC.

Reply:
@stevesg [also mentioned some thoughts here](https://github.com/grafana/grafana/issues/62361#issuecomment-2133099465) about using namespace UIDs vs folder titles. One of the main arguments for using folder titles is that Prometheus includes the filename in the same metric. I guess it's also a bit more recognizable from the user's perspective?

@stevesg what do you think?